docker/Dockerfile

```diff
@@ -1,88 +1,71 @@
-# - libxml2-dev/libxslt-dev for nokogiri, at least
-# - postgres-dev for pg/activerecord gems
-# - git for installing gems referred to use as git:// uri
-# - yarn for js dependency management
-#
-# runtime dependencies:
-# - file: for paperclip file type spoofing check
-# - nodejs: for ExecJS and asset compilation
-# - runit for process management (because we have multiple services)
-# - libreoffice: for pdf conversion
-# - ttf-freefont: needed for libreoffice
-# - redis: for backend key-value store
-# - postgresql-client - only needed for database dump
-#
-RUN apk --no-cache add --virtual build-dependencies \
-                    build-base \
-                    libxml2-dev \
-                    libxslt-dev \
-                    postgresql-dev \
-                    git \
-                    yarn \
-                    gmp=6.2.1-r2 \
-&& apk --no-cache add \
-                  file \
-                  nodejs \
-                  linux-headers \
-                  runit \
-                  ttf-freefont \
-                  libreoffice \
-                  redis \
-                  postgresql-client
-
-# add non-root user and group with alpine first available uid, 1000
-RUN addgroup -g 1000 -S appgroup \
-&& adduser -u 1000 -S appuser -G appgroup
+ENV APP_HOME /app
+WORKDIR $APP_HOME
 
-# create app directory in conventional, existing dir /usr/src
-RUN mkdir -p /usr/src/app && mkdir -p /usr/src/app/tmp
-WORKDIR /usr/src/app
+RUN apk --update-cache upgrade \
+ && apk --no-cache add \
+                   build-base \
+                   git \
+                   gmp=6.2.1-r2 \
+                   postgresql-dev \
+                   yarn
 
-######################
-# DEPENDENCIES START #
-######################
-# Env vars needed for dependency install and asset precompilation?? maybe not
 ENV RAILS_ENV production
 
-COPY Gemfile* ./
+COPY Gemfile Gemfile.lock $APP_HOME/
 
-# only install production dependencies,
-# build nokogiri using libxml2-dev, libxslt-dev
-# note: installs bundler version used in Gemfile.lock
-#
 RUN gem install bundler -v $(cat Gemfile.lock | tail -1 | tr -d " ") \
 && bundle config without test development devunicorn \
-&& bundle config build.nokogiri --use-system-libraries \
-&& bundle install
+&& bundle install --jobs 4 --retry 3 \
+&& rm -rf /usr/local/bundle/cache/*.gem \
+&& find /usr/local/bundle/gems/ -name "*.c" -delete \
+&& find /usr/local/bundle/gems/ -name "*.o" -delete
 
-COPY package.json yarn.lock ./
+COPY package.json yarn.lock $APP_HOME/
 RUN yarn install --frozen-lockfile --production --silent
 
-####################
-# DEPENDENCIES END #
-####################
-COPY . .
+COPY . $APP_HOME/
 
-# precompile assets, silently (as verbose)
 RUN SECRET_KEY_BASE=a-real-secret-key-is-not-needed-here \
 RAILS_ENV=production \
 bundle exec rails assets:precompile 2> /dev/null
 
-# tidy up installation
-RUN apk --update-cache upgrade && apk del build-dependencies
+RUN rm -rf node_modules
```

Comment:
Is it possible to use a `.dockerignore` file to avoid these files, as well as the files in `/usr/local/bundle/gems/` above, being copied rather than deleting them?

I see that `.dockerignore` is itself listed in `.dockerignore` and this makes sense for a one step process. Perhaps it can be removed so that it is copied into the stage 1 image and then added with:

```
RUN echo .dockerimage >> .dockerimage
```

so that it is not copied into the stage 2 image.

Reply:
Sadly the `.dockerignore` file only works on the first step of a build, copying files from the filesystem into the image. The`.dockerignore` has no affect when files are copied between images on subsequent steps 😢. Some discussion of it here: https://github.com/moby/moby/issues/33923.